pkg/kubelet/secret/secret_manager.go

```diff
@@ -0,0 +1,256 @@
+		if err != nil {
+			glog.Warningf("Unable to retrieve a secret %s/%s: %v", key.namespace, key.name, err)
+		}
+		results = append(results, result{secret: secret, err: err})
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for i, key := range keys {
+		secret := results[i].secret
+		err := results[i].err
+		if err != nil && !apierrors.IsNotFound(err) {
+			// If we couldn't retrieve a secret and it wasn't 404 error, skip updating.
+			continue
+		}
+		if item, ok := s.items[key]; ok {
+			if secret != nil && item.secret != nil {
+				// If the fetched version is not newer than the current one (such races are
+				// possible), then skip update.
+				newVersion, _ := storageetcd.Versioner.ObjectResourceVersion(secret)
+				oldVersion, _ := storageetcd.Versioner.ObjectResourceVersion(item.secret)
+				if newVersion <= oldVersion {
+					continue
+				}
+			}
+			item.secret = secret
+			item.err = err
+		}
+	}
+}
+
+// cachingSecretManager keeps a cache of all secrets necessary for registered pods.
+// It implements the following logic:
+// - whenever a pod is created or updated, the current versions of all its secrets
+//   are grabbed from apiserver and stored in local cache
+// - every GetSecret call is served from local cache
+// - every X seconds we are refreshing the local cache by grabbing current version
+//   of all registered secrets from apiserver
+type cachingSecretManager struct {
+	secretStore *secretStore
+
+	lock           sync.Mutex
+	registeredPods map[objectKey]*v1.Pod
+}
+
+func NewCachingSecretManager(kubeClient clientset.Interface) (Manager, error) {
+	csm := &cachingSecretManager{
+		secretStore:    newSecretStore(kubeClient),
+		registeredPods: make(map[objectKey]*v1.Pod),
+	}
+	go wait.NonSlidingUntil(func() { csm.secretStore.Refresh() }, time.Minute, wait.NeverStop)
+	return csm, nil
+}
+
+func (c *cachingSecretManager) GetSecret(namespace, name string) (*v1.Secret, error) {
+	return c.secretStore.Get(namespace, name)
+}
+
+// TODO: Before we will use secretManager in other places (e.g. for secret volumes)
+// we should update this function to also get secrets from those places.
+func getSecretNames(pod *v1.Pod) sets.String {
+	result := sets.NewString()
+	for _, reference := range pod.Spec.ImagePullSecrets {
+		result.Insert(reference.Name)
+	}
+	for i := range pod.Spec.Containers {
+		for _, envVar := range pod.Spec.Containers[i].Env {
+			if envVar.ValueFrom != nil && envVar.ValueFrom.SecretKeyRef != nil {
+				result.Insert(envVar.ValueFrom.SecretKeyRef.Name)
+			}
+		}
+	}
+	return result
+}
+
+func (c *cachingSecretManager) RegisterPod(pod *v1.Pod) {
+	for key := range getSecretNames(pod) {
+		c.secretStore.Add(pod.Namespace, key)
+	}
+	var prev *v1.Pod
+	func() {
+		key := objectKey{namespace: pod.Namespace, name: pod.Name}
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		prev = c.registeredPods[key]
+		c.registeredPods[key] = pod
+	}()
+	if prev != nil {
```

Comment:
are RegisterPod/UnregisterPod called from multiple goroutines? if so, the secret add/delete calls need to be in the lock, right? (otherwise an unregister followed by a register could race and end up deleting after adding)